### PR TITLE
Expose script members

### PR DESCRIPTION
**Pull request by JaimeBea**
- Members can now be exposed using the EXPOSE_MEMBERS and MEMBER macros
- Scripts have access to the owning GameObject via GetOwner()

### CODE REVIEW

**Review by jcarlos0305 — APPROVED** [review]

File: Game/Templates/Source [hunk 1]
```diff
@@ -1,5 +1,9 @@
 #include "{0}.h"
 
+EXPOSE_MEMBERS({0}) {1}
+
+{2};
+
 GENERATE_BODY_IMPL({0});
 
 void {0}::Start() {1}
```
Comment (FrancPS) [inline Game/Templates/Source:9]: For my own awareness in future reviews, here is how you updated the templates of a script file right?

(I mean, when creating a script, now this macro will be automatically generated in the .cpp)

File: Game/Assets/Prefabs/Test.prefab [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
+    "Root": {
+        "Name": "Test",
+        "Active": true,
+        "RootBoneName": "",
```
Comment (FrancPS) [inline Game/Assets/Prefabs/Test.prefab:5]: Does this need to be in develop?

File: Game/Assets/Scripts/ExposeTest.cpp [hunk 3]
```diff
@@ -0,0 +1,20 @@
+#include "ExposeTest.h"
```
Comment (FrancPS) [inline Game/Assets/Scripts/ExposeTest.cpp:1]: Does this file need to be in develop?

File: Project/Source/Modules/ModuleProject.cpp [hunk 4]
```diff
@@ -313,6 +317,19 @@ bool ModuleProject::Init() {
 	return true;
 }
 
+UpdateStatus ModuleProject::Update() {
+	for (ComponentScript& script : App->scene->scene->scriptComponents) {
+		if (App->time->HasGameStarted() && App->scene->sceneLoaded) {
```
Comment (FrancPS) [inline Project/Source/Modules/ModuleProject.cpp:322]: The if can go before the for, so we avoid the iteration if its false?

**Review by FrancPS — APPROVED** [review]
QA:

- I can't place a GameObject in the exposed GAME_OBJECT_UID member. I dont know where do I drop it from, but if it is from hierarchy, the inspector changes as soon as i click the GO i want!

- I'm missing a list of the member type names somewhere handy. Something like:
```
/* Member Types:
...
FLOAT
LONGLONG
...
GAME_OBJECT_UID 
...
*/
```
that generates automatically (in .cpp or .h) from the template.

- I also wanted to test if GAME_OBJECT_UID  would also work for components, but couldnt. Guess not.

**Comment by JaimeBea** [comment]
> QA:
> 
> * I can't place a GameObject in the exposed GAME_OBJECT_UID member. I dont know where do I drop it from, but if it is from hierarchy, the inspector changes as soon as i click the GO i want!
> * I'm missing a list of the member type names somewhere handy. Something like:
> 
> ```
> /* Member Types:
> ...
> FLOAT
> LONGLONG
> ...
> GAME_OBJECT_UID 
> ...
> */
> ```
> 
> that generates automatically (in .cpp or .h) from the template.
> 
> * I also wanted to test if GAME_OBJECT_UID  would also work for components, but couldnt. Guess not.

- I'll upload a fix for that.
- Components should be accessed via their owner GameObject.

File: Project/Source/Modules/ModuleProject.cpp [hunk 5]
```diff
@@ -313,6 +317,19 @@ bool ModuleProject::Init() {
...
+			Script* scriptInstance = script.GetScriptInstance();
+			if (scriptInstance != nullptr) {
+				scriptInstance->Update();
```
Comment (Jordi-Gil) [inline Project/Source/Modules/ModuleProject.cpp:325]: Why here instead of the Component Update?

**Review by Jordi-Gil — APPROVED** [review]

File: Project/Source/Modules/ModuleProject.cpp (hunk 5 above)
Comment (JaimeBea) [reply Project/Source/Modules/ModuleProject.cpp:325]: Because if we add a GameObject or Component in the middle of an iteration it crashes.

**Review by FrancPS — APPROVED** [review]

**Pull request merged**